### DISCUSSION TITLE
Add Documentation on how to test TaskFlow tasks

### DISCUSSION THREAD

**Discussion by 0xdholleran** [apache/airflow#25639] [General]
### What do you see as an issue?

I have spent the last few days trying to figure how to write a test to a TaskFlow task and still haven't figured it out, as it seems the set up necessary for testing this is not so trivial. It would be helpful if there existed some documentation on how write a test for testing a TaskFlow task

### Solving the problem

I think adding a testing section to the TaskFlow tutorial section of the documentation would be helpful. It could provide some example tests and a walkthrough of the components necessary for the setup of such tests

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Can you please provide more information on:


* what exactly you tried to test
* what was the problem
* what kind of documentation you missed
* example of similar documentation elsewhere that can show what you were trying to do and could not?

Without knowing what you were trying to do and failed, it's hard to come up with any documentation. in fact the best documentation that you can come up is one that user who wanted to do something and could not explains to other users how they finally achieved it, so you are likely the best person to make a contribution with similar kind of documentation.

If you will explain in detail what kind of tasks you wanted to do and how you failed/or what you found out during your trying, we will try to help you with the answers, but in exchange we will ask you to describe it in the way that a user like you would expect and search for. 

Maintainers are very bad in writing such documentation as they have too many assumptions, but if you explain what you wanted to do we can definitely guide you in finding the answers and then you could - as a way to contribute back - explain it as a Doc PR. Airlfow is created mostly by people like you (we have > 2100 contributors) who had some problems. solved it and contributed back.This is how FOSS works.

Happy to help you if you alaborate more to answer your questions, especially if you will help to submit such documentation back.

**Reply by josh-fell** [reply]
Should you find it helpful, you can always take a look at any of the unit tests in the repo as well for reference or to spark an idea for setting up your own tests.  [Here are unit tests specifically for TaskFlow decorators](https://github.com/apache/airflow/tree/main/tests/decorators).

**Reply by 0xdholleran** [reply]
@potiuk yeah I can create a PR to update the documentation once I figure out to setup the tests properly.

Basically I would like to better understand how to setup a test such that I can test that the output of a task using the `@task` decorator equals some value

For example, say I have the following dag in the file `example_dag.py`:

    from datetime import datetime
    from airflow.decorators import task
    from airflow import DAG
    
    
    with DAG(
        'example_dag',
        start_date=datetime(2022, 1, 1),
        schedule_interval=None
    ) as dag:
        @task
        def add_one(num: int) -> int:
        return num + 1
        result = add_one(5)

and I have the test file `example_dag_test.py`:

    from dags.example_dag import add_one


    def test_add_one_task():
        result = add_one(3)
        assert result == 4

this test fails with the error `AttributeError: 'int' object has no attribute 'operator'`. This seems to be because the result is still an XComArg object, and the dag needs to be run first before the result will be an integer right? The problem is I am not exactly sure how to simulate a dag run such that I can evaluate the output of a task

There is some documentation here on testing here [https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#testing-a-dag](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#testing-a-dag), which I have to tried to copy for setting up a dagrun and task instance, but it doesn't work

Suppose I try to modify the test to simulate a dagrun and taskinstance:

    from dags.example_dag import add_one, dag
    import datetime
    import pendulum
    from airflow.utils.state import DagRunState
    from airflow.utils.types import DagRunType
    
    DATA_INTERVAL_START = pendulum.datetime(2021, 9, 13, tz="UTC")
    DATA_INTERVAL_END = DATA_INTERVAL_START + datetime.timedelta(days=1)
    
    
    def test_add_one_task():
        dagrun = dag.create_dagrun(
            state=DagRunState.RUNNING,
            execution_date=DATA_INTERVAL_START,
            data_interval=(DATA_INTERVAL_START, DATA_INTERVAL_END),
            start_date=DATA_INTERVAL_END,
            run_type=DagRunType.MANUAL,
        )
        ti = dagrun.get_task_instance(task_id='add_one')
        ti.task = dag.get_task(task_id='add_one')
        ti.run(ignore_ti_state=True)
        assert ti.output == 6

This test fails with the error `AttributeError: 'NoneType' object has no attribute 'task'`. This is because the task instance is `None`. When I look at the dagrun object in the debugger, I can see that dagrun.task_instances equals the errors `Parent instance <DagRun at 0x7fa2dd2aa1c0> is not bound to a Session; lazy load operation of attribute 'task_instances' cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)`

Why does the dagrun need to be bound to a Session and how do I do that?

Basically, I feel like in order to run this test some background knowledge of how different components work is necessary which I do not know. If you could impart this knowledge to me, I can gladly update the documentation to try to make it more explicit

**Reply by potiuk** [reply]
If you want to use unit-testing, then I recommend you to do likel you usually do - i..e extract your "testable" methods into plain python metods (for example in a common package) and call it from within decorated function:

```
def my_testable_function(arg1, arg2):
     ...
     return result

@dag ...
@ task  ...
def my_task(dag_parameters):
   return my_testable_function(arg1= .., arg2=...).

```

This is absolutely standard and typical way you would approach testability of your python code. Then you can test "my_testable_function" using standard pytest tests.

I recommend this solution and I would encourage you to try it, see if it works for you and contribute some documentation back.

I think what you are trying to do is mocking "airflow framework" and basically trying to test airflow, not your own code. This is alraady done in Airflow - we have some 4000 tests running to test how framework works and how tasks interact with otehr task. But what you really want to test is whether your custom code does what is supposed to do - given an input, produce expected output. The solution I proposed does it.

Did I miss something?

**Comment by david-woelfle** [comment]
I find the solution @potiuk provided above too redundant for my taste and had a look in the airflow code. Seems like some of the Operator classes stores the actual decorated function as `function` argument. Hence for testing this:
```
@task
def dummy_task(arg):
    return arg + "_bar"
```

I do:
```
def test_dummy_task():
    actual_return = dummy_task.function("foo")
    expected_return = "foo_bar"
    assert actual_return == expected_return
```
which seems to work nicely.

**Reply by potiuk** [reply]
Nice one.
